qiskit/circuit/classicalfunction/boolean_expression.py

```diff
@@ -0,0 +1,122 @@
+        """
+        Args:
+            expression (str): The logical expression string.
+            name (str): Optional. Instruction gate name. Otherwise part of
+                        the expression is going to be used.
+        Raises:
+            MissingOptionalLibraryError: If tweedledum is not installed. Tweedledum is required.
+        """
+        if not HAS_TWEEDLEDUM:
+            raise MissingOptionalLibraryError(
+                libname='tweedledum',
+                name='BooleanExpression compiler',
+                pip_install='pip install tweedledum')
+        from tweedledum import BoolFunction
+        self._tweedledum_bool_expression = BoolFunction.from_expression(expression)
+        self.qregs = None  # TODO: Probably from self._tweedledum_bool_expression._signature
+
+        short_expr_for_name = (expression[:10] + '...') if len(expression) > 13 else expression
+        num_qubits = (self._tweedledum_bool_expression.num_outputs() +
+                      self._tweedledum_bool_expression.num_inputs())
+        super().__init__(name or short_expr_for_name, num_qubits=num_qubits, params=[])
+
+    def simulate(self, bitstring: str) -> bool:
+        """Evaluate the expression on a bitstring.
+
+        This evaluation is done classically.
+
+        Args:
+            bitstring: The bitstring for which to evaluate.
+
+        Returns:
+            bool: result of the evaluation.
+        """
+        from tweedledum import BitVec
+        bits = []
+        for bit in bitstring:
+            bits.append(BitVec(1, bit))
+        return bool(self._tweedledum_bool_expression.simulate(*bits))
+
+    def synth(self, registerless=True, synthesizer=None):
+        """Synthesis the logic network into a :class:`~qiskit.circuit.QuantumCircuit`.
+
+        Args:
+            registerless (bool): Default ``True``. If ``False`` uses the parameter names
+                to create registers with those names. Otherwise, creates a circuit with a flat
+                quantum register.
+            synthesizer (callable): A callable that takes self and returns a Tweedledum
+                circuit.
+        Returns:
+            QuantumCircuit: A circuit implementing the logic network.
+        """
+        if registerless:
+            qregs = None
+        else:
+            qregs = self.qregs
+
+        if synthesizer is None:
+            from tweedledum.synthesis import pkrm_synth  # pylint: disable=no-name-in-module
+            from .utils import tweedledum2qiskit
+            truth_table = self._tweedledum_bool_expression.truth_table(output_bit=0)
+            return tweedledum2qiskit(pkrm_synth(truth_table), name=self.name, qregs=qregs)
+        return synthesizer(self)
+
+    def _define(self):
+        """The definition of the boolean expression is its synthesis"""
+        self.definition = self.synth()
+
+    @classmethod
+    def from_dimacs_file(cls, filename: str):
```

Comment:
We could also use the same structure as in `PhaseOracle` to clean up the alternative constructor. But I'm also good with this one 🙂

Reply:
ok 👍🏻